libp2p/stream/bufferstream.nim

```diff
@@ -106,12 +114,23 @@ proc initBufferStream*(s: BufferStream,
   s.readReqs = initDeque[Future[void]]()
   s.dataReadEvent = newAsyncEvent()
   s.lock = newAsyncLock()
-  s.writeHandler = handler
+  s.writeLock = newAsyncLock()
   s.closeEvent = newAsyncEvent()
-  inc getBufferStreamTracker().opened
+  s.isClosed = false
+
+  if not(isNil(handler)):
+    s.writeHandler = proc (data: seq[byte]) {.async, gcsafe.} =
+      defer:
+        s.writeLock.release()
+      await s.writeLock.acquire()
```

Comment:
This will be a performance hit, we are running on a single scheduler on a single thread I'm not even sure we need this if code is written properly.
It's like putting a lock inside a loop.. Ideally should be 
```
lock()
while(!msgqueue.empty()) { send() }
unlock()
```

Reply:
Yes, I'm aware and that's unfortunate, but the problem is that we need to synchronize the  `closes` otherwise there is no way of emulating the half-closed behavior, this atleast from my experimentation - I'd love to be proven wrong.

Reply:
I'm already mentioned that `defer` construction is not working properly in Nim, please convert it tot `try/finally` construction which is working.

Reply:
> Yes, I'm aware and that's unfortunate, but the problem is that we need to synchronize the `closes` otherwise there is no way of emulating the half-closed behavior, this atleast from my experimentation - I'd love to be proven wrong.

Yeah I get your point, this here allows other code to be written "safely" altho I would just prefer other code be more aware :)

Reply:
> defer construction is not working properly in Nim, please convert it to try/finally construction which is working

How come? The compiler itself implements defer with a hidden try/finally.

Reply:
> > Yes, I'm aware and that's unfortunate, but the problem is that we need to synchronize the `closes` otherwise there is no way of emulating the half-closed behavior, this atleast from my experimentation - I'd love to be proven wrong.
> 
> Yeah I get your point, this here allows other code to be written "safely" altho I would just prefer other code be more aware :)

Actually, the issues I was seeing with synchronization, might be do to the weird `await` vs `result` behavior reported in #178. Using `result` as I do in #178 doesn't fail for me any more, so we might be able to remove this lock here.